magefile.go

```diff
@@ -0,0 +1,77 @@
+//go:build mage
+// +build mage
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+const BIN_DIR = "bin"
+
+var (
+	Default  = All
+	goexe    = "go"
+	gofmtexe = "gofmt"
+)
+
+func init() {
+	if exe := os.Getenv("GOEXE"); exe != "" {
+		goexe = exe
+		gofmtexe = "gofmt.exe"
+	}
+}
+
+func All() {
+	Generate()
+	Format()
+	Build()
+	Test()
+	Install()
+}
+
+func Generate() error {
+	fmt.Println("⚙️  Running generators...")
+	cmd := exec.Command(goexe, "run", filepath.Join(".", "function-metadata", "extract.go"))
+	cmd.Stdout = os.Stdout
+	return cmd.Run()
+}
+
+func Format() error {
+	fmt.Println("⚙️  Formatting files...")
+	cmd := exec.Command(gofmtexe, "-s", "-w", ".")
+	cmd.Stdout = os.Stdout
+	return cmd.Run()
+}
+
+func Build() error {
+	fmt.Println("Building...")
+	mdir := exec.Command("mkdir", filepath.Join(".", BIN_DIR))
+	cmd := exec.Command(goexe, "build", "-o", filepath.Join(".", BIN_DIR, "radd"))
+	mdir.Stdout = os.Stdout
+	cmd.Stdout = os.Stdout
+	mdir.Run()
+	return cmd.Run()
+}
+
+func Test() error {
+	fmt.Println("⚙️  Running tests...")
+	cmd := exec.Command(goexe, "test", filepath.FromSlash("./core/testing"))
+	cmd.Stdout = os.Stdout
+	return cmd.Run()
+}
+
+func Install() error {
+	fmt.Println("⚙️  Installing...")
+	cmd := exec.Command("cp", filepath.Join(".", BIN_DIR, "radd"), filepath.Join(os.Getenv("GOBIN"), "rad"))
```

Comment:
Unsure about that `cp` operation. On my machine, `$GOBIN` isn't defined so this just copies the `radd` binary to the repo's root. What's the intent? Is there a more robust approach?